### PR TITLE
INTMDB-420 cloud backup

### PR DESCRIPTION
**Pull request by rdharmedran**
## Description
 Feature update: cloud-provider-snapshots

Link to any related issue(s): 
https://jira.mongodb.org/browse/INTMDB-420
## Type of change:

- [✅]  Bug fix (non-breaking change which fixes an issue)
-[✅]  New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-[✅]  This change requires a documentation update

## Required Checklist:

-[✅]  I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added any necessary documentation (if appropriate)
- [ ] I have run `make fmt` and formatted my code

## Further comments
 - Renamed the module cloud-provider-snapshots  into cloud-backup
 - Added input parameter validation
 - Read function returned already deleted snapshot.So we have added condition.
- Delete function was not checking if snapshot is already exist before delete it.We have added condition for the same.
- Added cfn-test configurations.

### CODE REVIEW

File: cfn-resources/cloud-backup/cmd/resource/resource.go [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+	"github.com/mongodb/mongodbatlas-cloudformation-resources/util/validator"
+	log "github.com/sirupsen/logrus"
+	matlasClient "go.mongodb.org/atlas/mongodbatlas"
+)
+
+var CreateRequiredFields = []string{"ApiKeys.PublicKey", "ApiKeys.PrivateKey", "ClusterName", "ProjectId"}
+var ReadRequiredFields = []string{"ProjectId", "ClusterName", "Id", "ApiKeys.PublicKey", "ApiKeys.PrivateKey"}
+var UpdateRequiredFields []string
+var DeleteRequiredFields = []string{"ProjectId", "ClusterName", "Id", "ApiKeys.PublicKey", "ApiKeys.PrivateKey"}
+var ListRequiredFields = []string{"ApiKeys.PublicKey", "ApiKeys.PrivateKey", "ClusterName", "ProjectId"}
```
Comment (matt-condon) [inline cfn-resources/cloud-backup/cmd/resource/resource.go:18]: Can we consolidate this with what you have in [validator.go](cfn-resources/cloud-backup/cmd/validator_def/validator.go) (seems to be repeated logic)

File: cfn-resources/util/validator/validator.go [hunk 2]
```diff
@@ -18,22 +17,7 @@ type ValidatorDefinition interface {
 	GetListFields() []string
 }
```
Comment (matt-condon) [inline cfn-resources/util/validator/validator.go:18]: Is this interface still in use? Can it be removed?

File: cfn-resources/util/validator/validator.go [hunk 3]
```diff
@@ -18,22 +17,7 @@ type ValidatorDefinition interface {
...
-		fields = def.GetCreateFields()
-	case constants.Read:
-		fields = def.GetReadFields()
-	case constants.Update:
-		fields = def.GetUpdateFields()
-	case constants.Delete:
-		fields = def.GetDeleteFields()
-	default:
-		fields = def.GetListFields()
-	}
```
Comment (matt-condon) [inline cfn-resources/util/validator/validator.go:36]: Just wondering what the reason is for this refactor?

File: cfn-resources/cloud-backup/cmd/resource/resource.go (hunk 1 above)
Comment (rdharmedran) [reply cfn-resources/cloud-backup/cmd/resource/resource.go:18]: We have different fields required for each methods in different resources.
For Example,

Required Fields for Read Method in Cloud-Backup: {**"ProjectId", "ClusterName"**, "Id", "ApiKeys.PublicKey", "ApiKeys.PrivateKey"}
Required Fields for Read Method in Database-User: {"ApiKeys.PublicKey", "ApiKeys.PrivateKey", **"ProjectId", "DatabaseName", "Username"**}

File: cfn-resources/util/validator/validator.go (hunk 3 above)
Comment (rdharmedran) [reply cfn-resources/util/validator/validator.go:36]: To remove the Validator class from every module([cfn-resources/**<<module_name>>/cmd/validation/validate.go**]  and reuse the Validator class available in util package

File: cfn-resources/util/validator/validator.go (hunk 2 above)
Comment (rdharmedran) [reply cfn-resources/util/validator/validator.go]: Yes.We have removed this interface

**Review by matt-condon — APPROVED** [review]

**Pull request closed without merging**